src/OpenTelemetry.Api/Context/Propagation/TraceContextPropagator.cs

```diff
@@ -230,18 +237,72 @@ internal static bool TryExtractTracestate(string[] tracestateCollection, out str
+                List<string> listMembers = new List<string>();
+                for (int i = 0; i < tracestateCollection.Length; ++i)
+                {
+                    // test_tracestate_multiple_headers_different_keys
+                    listMembers.AddRange(tracestateCollection[i].Split(','));
+                }
+
+                if (listMembers.Count > 32)
+                {
+                    // https://github.com/w3c/trace-context/blob/master/spec/20-http_request_header_format.md#list
+                    // test_tracestate_member_count_limit
+                    return false;
+                }
 
-                // Iterate in reverse order because when call builder set the elements is added in the
-                // front of the list.
-                for (int i = tracestateCollection.Length - 1; i >= 0; i--)
+                var keySet = new HashSet<string>();
+                var result = new StringBuilder();
+                for (int i = 0; i < listMembers.Count; ++i)
                 {
-                    if (string.IsNullOrEmpty(tracestateCollection[i]))
+                    // https://github.com/w3c/trace-context/blob/master/spec/20-http_request_header_format.md#tracestate-header-field-values
+                    if (string.IsNullOrEmpty(listMembers[i]))
+                    {
+                        // Empty and whitespace - only list members are allowed.
+                        // Vendors MUST accept empty tracestate headers but SHOULD avoid sending them.
+                        continue;
+                    }
+
+                    // Spaces and horizontal tabs surrounding list-members are ignored.
+                    var listMember = listMembers[i].Trim(new char[] { ' ', '\t' });
+                    int keyLength = listMember.IndexOf('=');
+                    if (keyLength == listMember.Length || keyLength == -1)
                     {
+                        // Missing key or value in tracestate
                         return false;
                     }
 
-                    result.Append(tracestateCollection[i]);
+                    var key = listMember.Substring(0, keyLength);
+                    if (!ValidateKey(key))
+                    {
+                        // test_tracestate_key_illegal_characters in https://github.com/w3c/trace-context/blob/master/test/test.py
+                        // test_tracestate_key_length_limit
+                        // test_tracestate_key_illegal_vendor_format
+                        return false;
+                    }
+
+                    var value = listMember.Substring(keyLength + 1);
+                    if (!ValidateValue(value))
+                    {
+                        // test_tracestate_value_illegal_characters
+                        return false;
+                    }
+
+                    // ValidateKey() call above has ensured the key does not contain upper case letters.
+                    if (keySet.Contains(key))
```

Comment:
I think `Add` returns `false` if the item already exists. If so we could combine the `Contains` with the `Add`:
```csharp
if (!keySet.Add(key))
{
   // test_tracestate_duplicated_keys
   return false;
}
```

Reply:
Oh that's convenient. I'm adding it.